consensus/propagation/have_wants.go

```diff
@@ -63,58 +64,133 @@ func (blockProp *Reactor) handleHaves(peer p2p.ID, haves *proptypes.HaveParts) {
 		return
 	}
 
-	reqLimit := 1
-
-	// if enough requests have been made for the parts, don't request them.
-	for _, partIndex := range hc.GetTrueIndices() {
-		reqs := blockProp.countRequests(height, round, partIndex)
-		if len(reqs) >= reqLimit {
-			hc.SetIndex(partIndex, false)
-			// mark the part as fully requested.
-			fullReqs.SetIndex(partIndex, true)
-		}
-		// don't request the part from this peer if we've already requested it
-		// from them.
-		for _, p := range reqs {
-			// p == peer means we have already requested the part from this peer.
-			if p == peer {
-				hc.SetIndex(partIndex, false)
-			}
+	for _, index := range hc.GetTrueIndices() {
+		select {
+		case <-blockProp.ctx.Done():
+		case p.requestChan <- request{height: height, round: round, index: uint32(index)}:
 		}
 	}
 
-	if hc.IsEmpty() {
-		return
+	blockProp.broadcastHaves(haves, peer, int(parts.Total()))
```

Comment:
something to chew on is if we should be broadcasting this have if we haven't yet requested it

Reply:
good catch 👍 

i'll move the broadcast after sending the wants

Reply:
183c9c8f96034d3a030d266236b1fbbec4882a90

Reply:
The above is not fully functioning because the parts hashes are only for the original block and not the parity data. So, if we want to request parity data, that will fail.

I updated the implementation to keep track of haves and do some gymnastics to send also the parts hash: [9494f01](https://github.com/celestiaorg/celestia-core/pull/1773/commits/9494f0197edbe141e334d8b8b540faac1e02337c)

Reply:
> because the parts hashes are only for the original block and not the parity data

Wait what? the compact block has all part hashes and we generate proofs for original and parity data identically

Reply:
I based my reasoning on a comment but apparently is wrong. I'll revert and update the comment

Reply:
d631faef5c0c594c46352a03d7cc644fce1a157f